core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -538,32 +551,52 @@ class SnappySession(_sc: SparkContext) extends SparkSession(_sc) {
     }
   }
 
+
+
   private[sql] def clearWriteLockOnTable(): Unit = {
+    logInfo(s"cleanWriteLockOnTable for app ${sqlContext.sparkContext.appName}",
+      new Throwable("ClearWrite"))
     contextObjects.remove(SnappySession.BULKWRITE_LOCK) match {
       case null =>
       case lock => releaseLock(lock)
     }
   }
 
   private[sql] def grabLock(table: String, schemaName: String,
-      connProperties: ConnectionProperties): Any = {
+      connProperties: ConnectionProperties): Option[Any] = {
     SnappyContext.getClusterMode(sparkContext) match {
       case _: ThinClientConnectorMode =>
+        while (!SnappySession.executorAssigned) {
+          // Don't expect this case usually unless lots of applications are submitted in parallel
+          try {
+            Thread.sleep(100)
+          }
+          catch {
+            case _: InterruptedException => logWarning("Interrupted while waiting for executor.")
+          }
+          logDebug(s"grabLock waiting for resources to be " +
+            s"allocated ${SnappySession.executorAssigned}")
+        }
         val conn = ConnectionPool.getPoolConnection(table, connProperties.dialect,
           connProperties.poolProps, connProperties.connProps, connProperties.hikariCP)
         var locked = false
         var retrycount = 0
         do {
           try {
             logDebug(s" Going to take lock on server for table $table," +
-                s" current Thread ${Thread.currentThread().getId}")
-            val ps = conn.prepareCall(s"VALUES sys.ACQUIRE_REGION_LOCK(?)")
+                s" current Thread ${Thread.currentThread().getId} and " +
+              s"app ${sqlContext.sparkContext.appName}")
+
+            val ps = conn.prepareCall(s"VALUES sys.ACQUIRE_REGION_LOCK(?,?)")
             ps.setString(1, SnappySession.WRITE_LOCK_PREFIX + table)
+            ps.setInt(2, serializedWriteLockTimeOut)
             val rs = ps.executeQuery()
             rs.next()
             locked = rs.getBoolean(1)
             ps.close()
-            logDebug("Took lock on server. ")
+            logDebug(s"Took lock on server. for string " +
+              s"${SnappySession.WRITE_LOCK_PREFIX + table} and " +
+              s"app ${sqlContext.sparkContext.appName}")
           }
           catch {
             case e: Throwable =>
```

Comment:
How are you handling timeout? I mean if the timeout happens then some exception will come? Is it ok to eat the full Throwable or just handle just the timeout? Moreover on timeout are we retrying or should it fail?